internal/awslambda/extension/manager.go

```diff
@@ -0,0 +1,386 @@
+
+	wg.StartWithContext(ctx, func(c context.Context) {
+		err := server.Run(c)
+		// Shutting down due to context is an acceptable
+		// reason to shutdown and is not considered an error
+		if errors.Is(err, ctx.Err()) {
+			err = nil
+		}
+		errs <- err
+	})
+
+	select {
+	case err := <-errs:
+		// In the event that the lambda finished early before
+		// it had started accepting events without error,
+		// It is still considered an error since
+		// it was not correctly shutdown
+		if err == nil {
+			err = ErrServerEarlyExit
+		}
+		return multierr.Combine(err, m.initError(parent, err))
+	case <-clock.After(ctx, 100*time.Millisecond):
+		// In the event that statsd server returns an
+		// error during the allowed start up time
+		// the we can fail the registration
+	}
+
+	wg.StartWithContext(ctx, func(c context.Context) {
+		errs <- m.heartbeat(c, cancel)
+	})
+
+	wg.Wait()
+	close(errs)
+
+	var err error
+	for er := range errs {
+		err = multierr.Append(err, er)
+	}
+
+	if err != nil {
+		if errors.Is(err, ErrIssueProgress) {
+			return err
+		}
+		// Since context can be closed here, a seperate context is used
+		// to control sending exit data to the lambda
+		ctx, cancel := clock.TimeoutContext(context.Background(), time.Second)
+		defer cancel()
+
+		return m.exitError(ctx, err)
+	}
+
+	return nil
+}
+
+func (m *manager) heartbeat(ctx context.Context, cancel context.CancelFunc) error {
+	timer := clock.FromContext(ctx).NewTimer(100 * time.Millisecond)
+	defer timer.Stop()
+
+	for ctx.Err() == nil {
+		resp, err := m.nextEvent(ctx)
+		if err != nil {
+			return err
+		}
+
+		if resp.EventType == api.Shutdown {
+			m.log.WithField("reason", resp.ShutdownReason).Info("Shutting down extention handler")
+			break
+		}
+		m.log.WithFields(map[string]interface{}{
+			"request-id":     resp.RequestID,
+			"invocation-arn": resp.InvokedFunctionARN,
+			"deadline":       resp.Deadline,
+		}).Debug("Progressing further with the invocation")
+
+		deadline := time.Unix(0, resp.Deadline*int64(time.Millisecond))
+		waitTil := clock.FromContext(ctx).Until(deadline) / 2
+		timer.Reset(waitTil)
+
+		// To ensure we are not constantly sending heart beats to lambda,
+		// The timer will ensure that pause will be roughly half the time to the deadline
+		// The lambda will be continuely ready to accept data regardless
+		<-timer.C
+	}
+
+	cancel()
+
+	return nil
+}
+
+func (m *manager) register(ctx context.Context) error {
+	var (
+		buf bytes.Buffer
+		enc = json.NewEncoder(&buf)
+	)
+
+	enc.SetEscapeHTML(true)
+
+	err := enc.Encode(&api.RegisterRequestPayload{
+		Events: []api.Event{api.Invoke, api.Shutdown},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, api.RegisterEndpoint.GetUrl(m.domain), &buf)
+	if err != nil {
+		return err
+	}
+	req.Header.Set(api.LambdaExtensionNameHeaderKey, m.name)
+
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
```

Comment:
As a general pattern, if deferring like this it should defer as...
```
defer func() {
	_, _ = io.ReadAll(resp.Body) // or io.Copy, whatever floats your boat.
	_ = resp.Body.Close()
}
```

Reply:
I remember reading about the reasoning for this somewhere but can't find a good source, why do we drain the body before we close?